File: cost.py
import numpy as np 
import matplotlib.pyplot as plt 
from scipy import interpolate
from uav import model as model 

class cost:
    def __init__(self):
        self.m=model()
        self.beta=100
        self.tt=np.linspace(0,1,102)
        self.alpha=1
        self.ld=0.5
        
    def get_cost(self,sol1):
        self.x=sol1['x']
        self.y=sol1['y'] 
        self.k=self.x.size + 2
        self.TS=np.linspace(0,1,self.k)
        self.XS=np.insert(self.x,0,self.m.xs)
        self.XS=np.append(self.XS,self.m.xt)
        # self.YS=np.insert(self.y,[0,self.y.size],[self.m.ys,self.m.yt])
        self.YS=np.insert(self.y,0,self.m.ys)
        self.YS=np.append(self.YS,self.m.yt)

        xmax=self.XS[np.argmax(self.XS)]
        xmin=self.XS[np.argmin(self.XS)]
        # self.tt=np.linspace(xmin,xmax,100)
        xpts=interpolate.splrep(self.TS,self.XS)
        ypts=interpolate.splrep(self.TS,self.YS)
        # self.newXS=np.sort(self.XS)
        # self.newXS,self.newYS = zip(*sorted(zip(self.XS,self.YS)))

        # newpts=interpolate.splrep(self.newXS,self.newYS)
        self.xx=interpolate.splev(self.tt,xpts) 
        self.yy=interpolate.splev(self.tt,ypts)

        xx1=np.array([self.m.xs,self.xx[16],self.xx[33],self.xx[50],self.xx[67],self.xx[84],self.m.xt])
        yy1=np.array([self.m.ys,self.yy[16],self.yy[33],self.yy[50],self.yy[67],self.yy[84],self.m.yt])
        self.dx=np.diff(self.xx)
        self.dy=np.diff(self.yy)
        dx1=np.diff(self.xx[0:17])
        dx2=np.diff(self.xx[17:34])
        dx3=np.diff(self.xx[34:51])
        dx4=np.diff(self.xx[51:68])
        dx5=np.diff(self.xx[68:85])
        dx6=np.diff(self.xx[85:102])

        dy1=np.diff(self.yy[0:17])
        dy2=np.diff(self.yy[17:34])
        dy3=np.diff(self.yy[34:51])
        dy4=np.diff(self.yy[51:68])
        dy5=np.diff(self.yy[68:85])
        dy6=np.diff(self.yy[85:102])

        self.L=np.sum(np.sqrt(np.square(self.dx)+np.square(self.dy)))
        l1=np.sum(np.sqrt(np.square(dx1)+np.square(dy1)))
        l2=np.sum(np.sqrt(np.square(dx2)+np.square(dy2)))
        l3=np.sum(np.sqrt(np.square(dx3)+np.square(dy3)))
        l4=np.sum(np.sqrt(np.square(dx4)+np.square(dy4)))
        l5=np.sum(np.sqrt(np.square(dx5)+np.square(dy5)))
        l6=np.sum(np.sqrt(np.square(dx6)+np.square(dy6)))
        length=np.array([l1,l2,l3,l4,l5,l6])
        cost=(1-self.ld)*self.L 

        nobs=self.m.xobs.size 
        self.violation=0
        self.feasible=1
        
        for j in range(2,8,1):

            newx=np.linspace(xx1[j-2],xx1[j-1],7)
            newy=np.linspace(yy1[j-2],yy1[j-1],7)
            sum1=0
            flag=0

            for i in range(nobs):
                d1=np.square(np.square(newx[1]-self.m.xobs[i])+np.square(newy[1]-self.m.yobs[i]))
                d2=np.square(np.square(newx[2]-self.m.xobs[i])+np.square(newy[2]-self.m.yobs[i]))
                d3=np.square(np.square(newx[3]-self.m.xobs[i])+np.square(newy[3]-self.m.yobs[i]))
                d4=np.square(np.square(newx[4]-self.m.xobs[i])+np.square(newy[4]-self.m.yobs[i]))
                d5=np.square(np.square(newx[5]-self.m.xobs[i])+np.square(newy[5]-self.m.yobs[i]))

                temp=((1/d1)+(1/d2)+(1/d3)+(1/d4)+(1/d5))*self.m.tobs[i]
                sum1+=temp

                if d1<pow(self.m.robs[i],4):
                    flag+=1
                if d2<pow(self.m.robs[i],4):
                    flag+=1
                if d3<pow(self.m.robs[i],4):
                    flag+=1
                if d4<pow(self.m.robs[i],4):
                    flag+=1
                if d5<pow(self.m.robs[i],4):
                    flag+=1
            if flag>0:
                cost+=(self.ld)*(pow(length[j-2],2))*(sum1/5.00)


                # temp=1-(d/(self.m.robs[i]))

                # v=np.maximum(temp,np.zeros(1))
                # self.violation+=np.mean(v)
            
            # v=temp[np.argmax(temp)]
            # if v == 1:
            #     self.feasible=0
            #     self.violation+=1
        # if self.violation==0:
        #     self.feasible=1
        # else:
        #     self.feasible=0   
            
        
        # print(self.violation)
        
        # self.z=self.alpha*self.L+self.beta*self.violation*self.L
        # self.z=self.L*self.violation 
        # print(self.z)
        # self.z=np.random.randint(0,100)
        self.feasible=1
        self.z=cost 
        return self.z, self.feasible
        

            
                

File: COST.m
function [z,sol] = COST(sol1,model)
    %beta=100;
    x=sol1.x;
    y=sol1.y;
    
    xs=model.xs;
    ys=model.ys;
    xt=model.xt;
    yt=model.yt;
    tobs=model.tobs;
    xobs=model.xobs;
    yobs=model.yobs;
    robs=model.robs;
    lambda=0.5;
    
    
    XS=[xs x xt];
    YS=[ys y yt];

    k=numel(XS);
    TS=linspace(0,1,k);
    
    tt=linspace(0,1,102);
    xx=spline(TS,XS,tt);
    yy=spline(TS,YS,tt);
    %tt1=linspace(0,1,5);
    %xx1=spline(TS,XS,tt1);
    %yy1=spline(TS,YS,tt1);

    xx1=[xs,xx(17),xx(34),xx(51),xx(68),xx(85),xt];
    yy1=[ys,yy(17),yy(34),yy(51),yy(68),yy(85),yt];
    dx=diff(xx);
    dy=diff(yy);
    dx1=diff(xx(1:17));
    dx2=diff(xx(17:34));
    dx3=diff(xx(34:51));
    dx4=diff(xx(51:68));
    dx5=diff(xx(68:85));
    dx6=diff(xx(85:102));

    dy1=diff(yy(1:17));
    dy2=diff(yy(17:34));
    dy3=diff(yy(34:51));
    dy4=diff(yy(51:68));
    dy5=diff(yy(68:85));
    dy6=diff(yy(85:102));
    
    L=sum(sqrt(dx.^2+dy.^2));
    l1=sum(sqrt(dx1.^2+dy1.^2));
    l2=sum(sqrt(dx2.^2+dy2.^2));
    l3=sum(sqrt(dx3.^2+dy3.^2));
    l4=sum(sqrt(dx4.^2+dy4.^2));
    l5=sum(sqrt(dx5.^2+dy5.^2));
    l6=sum(sqrt(dx6.^2+dy6.^2));
    length=[l1,l2,l3,l4,l5,l6];
    nobs = numel(xobs); 
    Violation = 0;
    cost=(1-lambda)*L;
    
    for j=2:7

        newx=linspace(xx1(j-1),xx1(j),7);
        newy=linspace(yy1(j-1),yy1(j),7);
        sum1=0;
        flag=0;


        

        for k=1:nobs
            
            d1=(((newx(2)-xobs(k)).^2+(newy(2)-yobs(k)).^2)^2);
            d2=(((newx(3)-xobs(k)).^2+(newy(3)-yobs(k)).^2)^2);
            d3=(((newx(4)-xobs(k)).^2+(newy(4)-yobs(k)).^2)^2);
            d4=(((newx(5)-xobs(k)).^2+(newy(5)-yobs(k)).^2)^2);
            d5=(((newx(6)-xobs(k)).^2+(newy(6)-yobs(k)).^2)^2);
            temp=((1/d1)+(1/d2)+(1/d3)+(1/d4)+(1/d5))*tobs(k);
            sum1=sum1+temp;
            if d1<robs(k)^4
                flag=flag+1;
            end
            if d2<robs(k)^4
                flag=flag+1;
            end
            if d3<robs(k)^4
                flag=flag+1;
            end
            if d4<robs(k)^4
                flag=flag+1;
            end
            if d5<robs(k)^4
                flag=flag+1;
            end


            %v=max(1-d/robs(k),0);
            %Violation=Violation+mean(v);
        end

        if flag>0

            cost=cost+(lambda)*(length(j-1).^2)*(sum1/5.00);
        end

    end
    
    sol.TS=TS;
    sol.XS=XS;
    sol.YS=YS;
    sol.tt=tt;
    sol.xx=xx;
    sol.yy=yy;
    sol.dx=dx;
    sol.dy=dy;
    sol.L=L;
    sol.Violation=0;
    sol.IsFeasible=(Violation==0);
    
    %z=sol.L*(1+beta*sol.Violation);
    z=cost;
    
end